File: RxTango/python/examples/monitor_attribute.py
"""Subscribe to Tango CHANGE events — push Observable.

Prints each value as it arrives from the device event system.  Unlike polling,
no interval is needed: the device pushes values whenever the attribute changes.

.. note::
    Tango event subscriptions require a properly configured event system
    (zmq ports, event heartbeat from the device server).  This example is NOT
    part of the live demo but is included to show the API.

Usage:
    python monitor_attribute.py [device] [attribute] [event-type]

    device      defaults to tango://localhost:10000/sys/tg_test/1
    attribute   defaults to double_scalar
    event-type  defaults to periodic   (change | periodic | archive)
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from reactivex.scheduler.eventloop import AsyncIOScheduler

from rxtango import monitor_attribute


async def main() -> None:
    device     = sys.argv[1] if len(sys.argv) > 1 else "tango://localhost:10000/sys/tg_test/1"
    attr       = sys.argv[2] if len(sys.argv) > 2 else "double_scalar"
    event_type = sys.argv[3] if len(sys.argv) > 3 else "periodic"

    loop      = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)

    print(f"Monitoring {device}/{attr} ({event_type} events)  (Ctrl+C to stop)\n")

    monitor_attribute(device, attr, event=event_type).subscribe(
        on_next=lambda v: print(f"  {v}"),
        on_error=lambda e: print(f"  ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()  # run until Ctrl+C


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n  stopped.")

File: RxTango/cpp/examples/monitor_attribute.cpp
/**
 * Monitor a Tango attribute using push events — the multi-value, push-based pattern.
 *
 * Mirrors Python's monitor_attribute.py and Java's MonitorAttribute.java.
 * Uses RxTangoAttributeChangePublisher in Java; monitor_attribute<T> here.
 *
 * Note: Tango events require a properly configured event system (ZMQ ports
 * reachable from the client).  Use poll_attribute for polling instead.
 *
 * Usage:
 *   ./monitor_attribute [device] [attribute] [event-type]
 *   defaults: sys/tg_test/1  double_scalar  periodic
 */

#include <csignal>
#include <iostream>
#include <string>

#include <rxtango/rxtango.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string device = argc > 1 ? argv[1] : "tango://localhost:10000/sys/tg_test/1";
    const std::string attr   = argc > 2 ? argv[2] : "double_scalar";
    const std::string event  = argc > 3 ? argv[3] : "periodic";

    std::signal(SIGINT, [](int) { g_running = 0; });

    std::cout << "Monitoring " << device << "/" << attr
              << " [" << event << " events]  (Ctrl+C to stop)\n\n";

    auto sub = rxtango::monitor_attribute<double>(device, attr, event)
        .subscribe(
            [](double v) { std::cout << "  event: " << v << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();   // → unsubscribe_event() called; stream torn down cleanly
    std::cout << "\n  stopped.\n";
    return 0;
}
